[C팀] 백엔드파트 코드리뷰용 PR

anxi01 opened the pull request:
개발파트 심사위원 코드리뷰용 PR입니다.

바쁜 와중에 코드 리뷰해 주셔서 정말 감사드립니다.

HoyeongJeon committed:
feat: (#65) QuickStartActivitySelectRequest를 구현한다.

HoyeongJeon committed:
feat: (#65) QuickStartActivitySelectController와 스웨거를 구현한다.

HoyeongJeon committed:
feat: (#65) QuickStartActivitySelectService를 생성한다.

HoyeongJeon committed:
feat: (#65) QuickStartActivitySelectSwagger를 만들고 Controller를 구현한다.

HoyeongJeon committed:
feat: (#65) QuickStartActivitySelectRequest를 구현한다.

HoyeongJeon committed:
feat: (#65) QuickStartActivitySelectService를 구현한다.

HoyeongJeon committed:
feat: (#65) QuickStartActivitySelectServiceTest를 작성한다.

devxb commented on Dockerfile-local:
```diff
@@ -0,0 +1,17 @@
+
+EXPOSE 8080
+ENTRYPOINT ["java"]
+CMD ["-jar", "backend.jar"]
```
사소한것이지만 파일의 마지막에 개행을 넣으시면 아래 사진과 같은것을 안 볼 수 있습니다!
![스크린샷 2024-11-06 오후 10 16 22](https://github.com/user-attachments/assets/76736e16-9312-43cd-aefd-6d0cdbbd34d5)

HoyeongJeon replied:
감사합니다!!

devxb commented on build.gradle:
```diff
@@ -24,14 +24,53 @@ repositories {
-	implementation 'org.springframework.boot:spring-boot-starter-validation'
+	// WEB
 	implementation 'org.springframework.boot:spring-boot-starter-web'
-	compileOnly 'org.projectlombok:lombok'
-	developmentOnly 'org.springframework.boot:spring-boot-devtools'
-	annotationProcessor 'org.projectlombok:lombok'
+
+	// Testing
 	testImplementation 'org.springframework.boot:spring-boot-starter-test'
+	testImplementation 'org.springframework.security:spring-security-test'
 	testRuntimeOnly 'org.junit.platform:junit-platform-launcher'
+	testImplementation 'com.h2database:h2'
+
+	// Development tools (for local development only)
+	developmentOnly 'org.springframework.boot:spring-boot-devtools'
+
+	// DB, JPA
+	runtimeOnly 'com.mysql:mysql-connector-j'
+	implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
+
+	// Lombok (code simplification)
+	compileOnly 'org.projectlombok:lombok'
+	annotationProcessor 'org.projectlombok:lombok'
+
+	// Validation (for request/response validation)
+	implementation 'org.springframework.boot:spring-boot-starter-validation'
+
+	// WebSocket (for real-time communication)
+	implementation 'org.springframework.boot:spring-boot-starter-websocket'
+
+	// Swagger-UI
+	implementation 'org.springdoc:springdoc-openapi-starter-webmvc-ui:2.0.2'
+
+	// JWT
+	implementation 'io.jsonwebtoken:jjwt-api:0.12.6'
+	runtimeOnly 'io.jsonwebtoken:jjwt-impl:0.12.6'
+	runtimeOnly 'io.jsonwebtoken:jjwt-jackson:0.12.6'
```
마찬가지로 사소한것이지만, 모든 설정을 하나의 gradle파일에 두면, 프로젝트가 커졋을때 관리하기 힘들어집니다.

아래 사진과 같이, 역할별로 gradle파일을 만들고 (메인)build.gradle에 apply를 해주면 관리하기 편해지니 시도해보셔도 좋을거 같습니다.
[참고용 레포지토리도 링크로 남겨드립니다](https://github.com/git-goods/gitanimals/blob/main/build.gradle)

![스크린샷 2024-11-06 오후 10 18 07](https://github.com/user-attachments/assets/63e3e0ee-3076-46a8-9da6-a65f4f157af6)

<build.gradle>
apply from: "gradle/db.gradle"
apply from: "gradle/jwt.gradle"
apply from: "gradle/core.gradle"
apply from: "gradle/test.gradle"
apply from: "gradle/slack.gradle"
apply from: "gradle/spring.gradle"
apply from: "gradle/monitor.gradle"
apply from: "gradle/jetbrains.gradle"

anxi01 replied:
오 적용해보겠습니다! 감사합니다!

HoyeongJeon replied:
좋은 팁 감사합니다!!

devxb commented on build.gradle:
```diff
@@ -24,14 +24,53 @@ repositories {
+	implementation 'org.springframework.boot:spring-boot-starter-data-redis'
+
+	// WebFlux 외부 API 호출
+	implementation 'org.springframework.boot:spring-boot-starter-webflux'
```
mvc와 webflux를 함께 쓰시나요?
- 이러면 서버는 Tomcat으로 뜨게될까요 Netty로 뜨게될까요?
- mvc와 webflux는 어떤상황에 같이 사용할 수 있을까요?

HoyeongJeon replied:
현재 저희 서비스는 WebFlux를 활성화하지 않았기에, Tomcat으로 서버가 구동됩니다. 
WebFlux를 추가한 이유는 외부 API(OpenAI, ClovaStudio)를 사용하려고 했기에 추가했습니다!

devxb commented on src/main/java/spring/backend/activity/application/CreateQuickStartService.java:
```diff
@@ -0,0 +1,48 @@
+@Service
+@RequiredArgsConstructor
+@Log4j2
+@Transactional
```
저도 DDD를 좋아하는데, 저는 트랜잭션 관리의 용이성 때문에 domain Service는 domain레이어에 구현하는걸 추천드립니다. (자연스럽게 `@Transactional`도 domain 밖으로 빠져나오지 않겠죠?)
application 레이어에서는 여러 domain service 를 조합해서 사용해야 하거나 I/O 작업이 발생하는데, 트랜잭션 안에는 넣으면 안되는 상황에 사용합니다.

anxi01 replied:
- `application service` : 여러개의 도메인 서비스의 오케스트레이션 및 트랜잭션분리
- `domain service` : 각 도메인 특정한 로직을 담당

devxb commented on src/main/java/spring/backend/activity/domain/value/Keyword.java:
```diff
@@ -0,0 +1,37 @@
+import jakarta.persistence.Enumerated;
+import lombok.*;
+
+@Embeddable
```
클린아키텍처를 사용한것은 아닌가요?
domain layer가 POJO로 되어있고 의존성도 application -> domain <- infra 로 가고있는거 같아서 클린아키텍처 + DDD라고 생각했는데 여기 jakarta쪽 어노테이션이 있군요

anxi01 replied:
클린아키텍처를 사용하려고 하였으나, 잘 모르는 상황이었어서 사용하지 않았습니다..!
그대신 표현 -> 응용 -> 도메인 <- 인프라 4계층으로 DDD를 설계하려고 하였습니다!

도메인 계층은 POJO로 구현하고 인프라에서 JPA를 사용하였지만, 
밸류 타입의 경우 도메인 계층에서 jakarta 어노테이션을 사용하여 POJO로 생성하지 못했습니다..

밸류 타입도 도메인은 POJO로 변경하고, 인프라에 jakarta 어노테이션 사용 및 mapper를 통해 관리하도록 변경하겠습니다!
감사합니다~

devxb commented on src/main/java/spring/backend/activity/dto/request/QuickStartRequest.java:
```diff
@@ -0,0 +1,42 @@
+import jakarta.validation.constraints.*;
+import spring.backend.activity.domain.value.Type;
+
+public record QuickStartRequest(
```
저는 Request와 Response는 사용되는 layer에 위치시키는것을 선호하는데요.
이유는, 공용 Request, Response를 만들게되면, 해당 Request, Response가 변경되었을때 함께 변경해야하는 레이어가 많아지기 때문입니다. 그러다보면, Request, Resposne를 쉽게 변경할 수 없게되어요.

반대로, Request, Response가 controller에서만 사용된다면, controller layer하위에 패키지를 위치시키고 외부에서는 접근못하도록 정책을 세워두면 관리하기가 조금더 용이해질것같습니다.

HoyeongJeon replied:
조언 감사합니다!!!
저희 dto가 controller에서 응답을 받고 주는 용도도 있지만, application이나 response dto를 만들기 위한 세부 dto도 존재합니다.
조언해주신 것에 아이디어를 얻어 controller를 거치는지 아닌지에 따라 dto의 위치를 구분해보겠습니다!

devxb commented on src/main/java/spring/backend/activity/dto/request/QuickStartRequest.java:
```diff
@@ -0,0 +1,42 @@
+        @Min(value = 0, message = "시작 시간의 분은 최소 0이어야 합니다.")
+        @Max(value = 59, message = "시작 시간의 분은 최대 59이어야 합니다.")
+        @Schema(description = "시작 시간의 분", example = "30")
+        Integer minute,
```
Wrapping타입을 사용하시는 이유가 있나요?
최근에는 코틀린을 써서 요즘 자바는 다를 수 있겠지만, wrapping 타입을 사용하면 auto unboxing, boxing이 발생하면서 불필요한 부하가 발생하게 됩니다.

anxi01 replied:
값이 없을 경우를 대비해서 처음에 Wrapper 타입을 사용했습니다!
오토 박싱에 대한 성능 문제를 고려하지 않았어서 값이 null일 수 있을 때만 Wrapper 타입을 사용하고 이외 전부 primitive 타입을 사용하도록 변경하겠습니다!
감사합니다~

devxb commented on src/main/java/spring/backend/activity/infrastructure/persistence/jpa/entity/ActivityJpaEntity.java:
```diff
@@ -0,0 +1,45 @@
+import java.time.LocalDateTime;
+import java.util.UUID;
+
+@Entity
+@Table(name = "activity")
+@Getter
+@SuperBuilder
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+public class ActivityJpaEntity extends BaseLongIdEntity {
```
도메인에서 JPA를 분리하는게 어떤 장점이 있을까요?
실제로 JPA가 다른 기술로 변경되는일이 발생할까요? 오히려 기술을 위한 불필요한 코드(interface, entity, mapper...)가 늘어나는것은 아닐까요?

클린아키텍처에서 말하는 POJO도메인이 백엔드 개발자한테 어떤 장점을 가져다주는지 고민해보셔도 좋을거 같습니다.

anxi01 replied:
사실 이 부분에서 고민을 많이 했습니다.
저희는 최범균님의 `도메인 주도 개발 시작하기: DDD 핵심 개념 정리부터 구현까지`를 학습하여 프로젝트를 진행중인데요!
책에서도 그렇고 김영한님의 의견에서도 실용적인 부분도 고려해서 기술 자체가 변경되지 않으면 분리를 하지 않고 사용하시더라고요!
![스크린샷 2024-11-25 오후 7 29 46](https://github.com/user-attachments/assets/d9e123b2-0ffa-42c8-a5af-ce16d1c88fa3)

준영님의 말씀대로, 변경되는 일이 거의 없을 거라 생각이 들고 불필요한 코드가 늘어나는 것에 동의합니다!
하지만 기술에 독립적인 POJO로 개발함으로써, 변화에 유연하고 도메인 로직에 집중할 수 있는 장점이 있다고 생각합니다!
물론 기술이 바뀌지는 않겠지만, 이번에 처음 DDD를 공부하고 도입하는 과정에서 엄격하게 DDD를 사용해보고 싶어서 POJO 도메인으로 사용했습니다!

devxb commented on src/main/java/spring/backend/activity/infrastructure/persistence/jpa/dao/ActivityJpaDao.java:
```diff
@@ -0,0 +1,30 @@
+import org.springframework.data.jpa.repository.Query;
+import spring.backend.activity.dto.response.HomeActivityInfoResponse;
+import spring.backend.activity.infrastructure.persistence.jpa.entity.ActivityJpaEntity;
+import spring.backend.activity.query.dao.ActivityDao;
```
패키지명이 스프링에서 공식 레포지토리 인줄알았네요 ㅎㅎ 🤣

anxi01 replied:
ㅋㅋㅋㅋㅋㅋㅋㅋ 의도치 않게 공식 레포지토리처럼 네이밍이 되었네요 😆

devxb commented on src/main/java/spring/backend/activity/infrastructure/persistence/jpa/entity/ActivityJpaEntity.java:
```diff
@@ -0,0 +1,45 @@
+@Getter
+@SuperBuilder
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+public class ActivityJpaEntity extends BaseLongIdEntity {
```
DDD를 사용했다 들었는데, Activity와 QuickStart 모두 AggregateRoot 인건가요?
Activity에서 QuickStart를 id로 간접참조 하고 있는데, 같은도메인에서 직접참조를 하지 않으신 이유가 있는걸까요?

anxi01 replied:
네 맞습니다. 원래 Activity를 애그리거트로 설정하고 QuickStart는 엔티티로 사용하여 Activity 애그리거트 내에 Activity와 QuickStart를 두려고 했습니다 !
그렇지만 QuickStart가 Activity와 분리된 기능, 예를 들어 조회, 생성, 수정 등의 기능이 Activity와 독립적으로 동작할 필요가 있었습니다. 그래서 Activity : QuickStart = 1 : N 관계에서 QuickStart만 따로 다루는 것이 더 유리하다고 설계 당시 판단하였습니다.
또한 Activity에서 QuickStart를 ID로 간접 참조하게 된 이유는 QuickStart가 독립적으로 다루어질 수 있어야 했기 때문입니다. 직접 참조를 하게 되면 지연 로딩을 해야 하는데 애그리거트 로딩 전략을 통해 즉시 로딩으로 조회 시점에서 애그리거트를 완전한 상태가 되도록 책에서 언급했어서 그대로 따라했던 것 같습니다.. ㅎㅎ
물론 책에서 즉시 로딩이 반드시 정답은 아니라고 설명했기 때문에, 지연 로딩을 사용해 QuickStart만 로드하는 방식도 좋았을 것 같습니다. Activity와 QuickStart는 같은 애그리거트가 아닌 만큼 도메인 분리를 하는게 더 좋을 거 같네요!
이번 밋업 끝나고 리팩토링 하면서 직접 참조로 변경해보겠습니다. 감사합니다!

devxb commented on src/main/java/spring/backend/activity/presentation/CreateQuickStartController.java:
```diff
@@ -0,0 +1,29 @@
+    private final CreateQuickStartService createQuickStartService;
+
+    @Authorization
+    @PostMapping("/v1/quick-starts")
```
아래와 같이 request header로 버저닝하는 방법도 있습니다! (url 버저닝도 직관적이라 매우 좋은 방법이에요 :+1:)

@PostMapping(path = ["/gotchas"], headers = ["Api-Version=2"])

devxb commented on src/main/java/spring/backend/activity/presentation/FinishActivityController.java:
```diff
@@ -0,0 +1,27 @@
+    private final FinishActivityService finishActivityService;
+
+    @Authorization
+    @PatchMapping("/v1/activities/{activityId}/finish")
```
Patch, Post, Put등을 결정할때는 애매한경우에 API가 멱등한지의 여부로 판단하기도 합니다.
- Patch : API가 멱등하지 않을때 (여러번 호출되면 호출될때마다 다른응답이 나올 수 도 있을때)
- Put : API가 멱등할때

anxi01 replied:
API가 멱등한지의 여부도 잘 따져보아야겠네요..!
감사합니다 !

devxb commented on src/main/java/spring/backend/activity/presentation/swagger/CreateQuickStartSwagger.java:
```diff
@@ -0,0 +1,24 @@
+import spring.backend.member.domain.entity.Member;
+
+@Tag(name = "QuickStart", description = "빠른 시작")
+public interface CreateQuickStartSwagger {
+
+    @Operation(
+            summary = "빠른 시작 생성 API",
+            description = "빠른 시작을 생성합니다.",
+            operationId = "/v1/quick-starts"
+    )
+    @ApiErrorCode({GlobalErrorCode.class, QuickStartErrorCode.class})
+    ResponseEntity<RestResponse<Long>> createQuickStart(@Parameter(hidden = true) Member member, QuickStartRequest request);
+}
```
Controller와 Swagger를 분리했을때 어떤 이점이 있을까요?
실제로 변경이 발생하지 않을 수 있는데, 너무 이르게 추상화를 하게 되면 소스코드를 파악하기 힘들게 할 뿐더러 작성해야하는 코드량도 늘어나게 됩니다! (그런 의미에서 잘 사용하던 JPA가 다른것으로 변경될일은 매우매우 적기 때문에 클린아키텍처를 잘 사용하지 않아요)

잘 덜어내는것 또한 프로젝트의 퀄리티를 올리는데 중요한 것 같아요.

anxi01 replied:
Controller와 Swagger를 분리한 이유는 주로 운영 코드에 외부 코드의 침범을 최소화하려는 목적 하나뿐이었습니다..! 
이를 통해 API 문서와 실제 로직이 독립적으로 관리되므로 유지보수에 더 좋지 않을까 싶었습니다.

그런데 팀원의 얘기도 들어보니 분리하면서 코드가 분리되어 초반에는 헷갈렸다는 의견도 있었습니다.
잘 덜어냈어야하는데 더 복잡하게 작업을 하게 됐네요..
"잘 덜어내는 것"에 대해 더 고민해보겠습니다! 
감사합니다 :)

devxb commented on src/main/java/spring/backend/activity/presentation/UpdateQuickStartController.java:
```diff
@@ -0,0 +1,29 @@
+    private final UpdateQuickStartService updateQuickStartService;
+
+    @Authorization
+    @PatchMapping("/v1/quick-starts/{quickStartId}")
```
`@ResponseStatus(HttpStatus.OK)`

와 같은 어노테이션을 추가하면 성공 케이스에서 해당 API가 어떤 응답 코드를 내려주는지 조금 더 명시적으로 표현할 수 있습니다!

devxb commented on src/main/java/spring/backend/auth/application/HandleOAuthLoginService.java:
```diff
@@ -0,0 +1,59 @@
+        Member member = createMemberWithOAuthService.createMemberWithOAuth(createMemberWithOAuthRequest);
+
+        return new LoginResponse(jwtService.provideAccessToken(member), refreshTokenService.saveRefreshToken(member), member.getRole());
+    }
```
application service를 잘 사용해주신것 같아요 👍

anxi01 replied:
준영님 레포지토리 많이 참고했습니다! 감사합니다 ☺️

devxb commented on src/main/java/spring/backend/auth/application/HandleOAuthLoginService.java:
```diff
@@ -0,0 +1,59 @@
+
+    private final OAuthRestClientFactory oAuthRestClientFactory;
+
+    private final CreateMemberWithOAuthService createMemberWithOAuthService;
```
`auth.application` -> `member.application` 으로 의존성이 가고있는데요.
`auth.application`가 `member.application` 보다 낮은 위계일까요? (member.application이 변경되면 auth.application이 변경되어야 할까요?)

추가로, 서로 다른 도메인간의 통신은 어떤방식으로 하는게 좋을까요?

devxb commented on src/main/java/spring/backend/auth/application/OnboardingSignUpService.java:
```diff
@@ -0,0 +1,54 @@
+
+@Service
+@RequiredArgsConstructor
+@Transactional
+@Log4j2
+public class OnboardingSignUpService {
```
위에서도 말했지만, Transactional이 붙어있거나 도메인로직과 밀접한 domain service는 주로 domain layer에 정의합니다! (데이터 변경을 관리하기에 용이해요.)

anxi01 replied:
네! 도메인 로직은 도메인 서비스에 이관하고, 오케스트레이션할 때 응용 서비스로 생성하겠습니다!
감사합니다!

devxb commented on src/main/java/spring/backend/auth/application/RotateAccessTokenService.java:
```diff
@@ -0,0 +1,46 @@
+    private final RefreshTokenService refreshTokenService;
+
+    public RotateAccessTokenResponse rotateAccessToken(String refreshToken) {
+        UUID memberId = extractMemberIdFromRefreshToken(refreshToken);
```
Id는 선형적으로 증가하도록 하는게 DB index의 공간, 시간 활용도면에서 좋습니다.

랜덤한 ID를 생성하게 되면, (MySQL을 사용한다면) DB index 자료구조로 인해서, 저장 및 조회에 더 많은 overhead가 발생하게 되어요.

twitter snowflake id generator 를 키워드로 검색해보셔도 좋을거 같습니다.

HoyeongJeon replied:
조언 감사드립니다!
덕분에 UUID의 특성(랜덤한 ID 생성)이 DB Index의 공간 , 시간 활용도를 떨어뜨린다는 부분(인덱스 단편화 증가, I/O 증가, 캐시 효율성 감소) 등을 공부했습니다! 알려주신 twitter snowflake id generator를 공부해보고 이를 id로 적용할까 고민하였지만, 현재 많은 부분에서 member Id 타입이 uuid로 설정되어있고, twitter snowflake id generator를 사용하면 db 스키마도 변경해야하게 되니, qa를 진행하고 있는 현재 상황에선 twitter snowflake id보다 sequential UUID를 사용하는 것이 더 좋다고 판단하였습니다.
덕분에 많은 공부를 할 수 있었습니다.

devxb commented on src/main/java/spring/backend/auth/infrastructure/OAuthRestClientFactory.java:
```diff
@@ -0,0 +1,35 @@
+@RequiredArgsConstructor
+public class OAuthRestClientFactory {
+
+    private final GoogleOAuthRestClient googleOAuthRestClient;
+
+    private final NaverOAuthRestClient naverOAuthRestClient;
+
+    private final KakaoOAuthRestClient kakaoOAuthRestClient;
+
```
OAuth 추가는 충분히 발생할 수 있고, 벌써 3개의 OAuth client가 존재하니까 이런것은 좋은 추상화 인것 같습니다.

anxi01 replied:
감사합니다 😊

devxb commented on src/main/java/spring/backend/core/application/JwtService.java:
```diff
@@ -0,0 +1,105 @@
+
+
+@Service
+public class JwtService {
```
Jwt는 뭔가 auth 도메인에 있을거 같은 네이밍 이네요.
다른 도메인에서도 Jwt 암/복호화 를 해야해서 core에 정의되어 있는것이라면, auth로 옮기고 Filter나 Spring Request scope 등을 활용해서 auth에서 미리 암복호화를 할수도 있을거 같아요.

devxb commented on src/main/java/spring/backend/member/application/CreateMemberWithOAuthService.java:
```diff
@@ -0,0 +1,63 @@
+
+    private final MemberRepository memberRepository;
+
+    public Member createMemberWithOAuth(CreateMemberWithOAuthRequest request) {
```
함수가 조금 긴것같은데, 적절히 분리해주셔도 가독성 향상에 좋을거 같습니다.

devxb replied:
+Transactional을 붙여주는게 안정적이지 않을까요?

anxi01 replied:
함수 분리 및 @Transactional을 붙여 가독성 향상과 안정성을 높였습니다!
감사합니다!

devxb commented on src/main/java/spring/backend/member/application/ReadMemberHomeService.java:
```diff
@@ -0,0 +1,56 @@
+import org.springframework.transaction.annotation.Transactional;
+import spring.backend.activity.dto.response.HomeActivityInfoResponse;
+import spring.backend.activity.dto.response.QuickStartResponse;
+import spring.backend.activity.query.dao.ActivityDao;
+import spring.backend.activity.query.dao.QuickStartDao;
```
member -> activity로 의존성이 가고있네요 ㅎㅎ
DDD를 사용해서 도메인별로 로직을 잘 분리해주어도. 사용하는쪽의 의존성을 관리하지 않으면 관리하기 힘든 코드가 될 수 있습니다! (지금 그렇다는건 아니에요)

다른 도메인들끼리는 어떤 방식으로 호출을하는게 좋을지(네트워크 호출? 코드 호출?), 레이어는 어디끼리 참조하고 있어야 좋을지 고민해보셔도 좋을거 같아요. 
ex. member.application -> activity.domain?
ex. member.application -> activity.application?

devxb commented on src/test/java/spring/backend/activity/dto/request/QuickStartRequestTest.java:
```diff
@@ -0,0 +1,74 @@
+        @Test
+        @DisplayName("null 값일 경우 에러가 발생한다.")
+        void whenNameIsNull_thenValidationFails() {
+            QuickStartRequest request = new QuickStartRequest(null, 12, 30, "오전", 300, Type.OFFLINE);
+            Set<ConstraintViolation<QuickStartRequest>> violations = validator.validate(request);
+
+            assertThat(violations).isNotEmpty();
+            assertThat(violations).anyMatch(violation -> violation.getMessage().contains("이름은 필수 입력 항목입니다."));
```
given, when, then을 주석으로라도 명시해놓는게 추후 코드를 다시 읽거나 다른사람이 테스트 코드를 파악할때 도움이 됩니다!

anxi01 replied:
쓰다 안쓰다 했는데 확실하게 써주는게 좋겠네요! 
감사합니다 ㅎㅎ

devxb commented on src/main/java/spring/backend/recommendation/application/GetRecommendationsFromClovaService.java:
```diff
@@ -0,0 +1,49 @@
+    private static final Pattern CONTENT_PREFIX_PATTERN = Pattern.compile(".*content :");
+    private static final String LINE_SEPARATOR = "\n";
+    private final RecommendationProvider recommendationProvider;
+    public List<ClovaRecommendationResponse> getRecommendationsFromClova(ClovaRecommendationRequest clovaRecommendationRequest) {
```
요기 함수위에 개행이 빠져있는것 같아요 ㅎㅎ

devxb commented on src/main/java/spring/backend/recommendation/infrastructure/clova/application/ClovaService.java:
```diff
@@ -0,0 +1,42 @@
+@Service
+@RequiredArgsConstructor
+@Log4j2
+public class ClovaService {
```
Service라는 네이밍은 통상적으로 어떤 역할을할때 사용되고 있을지 고민해봐도 좋을거 같습니다.

HoyeongJeon replied:
Service 네이밍은 비즈니스 로직의 중심이 되는 계층이나 애플리케이션의 핵심 기능을 구현하고 관리하는 역할이라 알고있습니다.

코드를 작성하던 당시 추천 목록을 받아오니 핵심 기능을 구현한다 생각해 ClovaService라 지었습니다. 

현재 저희 애플리케이션의 메인 로직은 추천 목록을 받아오는 것이 아닌, 받아온 추천 목록을 파싱해 사용자에게 전달하는 것이기에 Service가 아닌 RecommendationProvider로 이름을 변경했습니다. 

더 깊게 생각할거리를 주셔서 감사합니다.

devxb commented on src/main/java/spring/backend/recommendation/infrastructure/clova/application/ClovaService.java:
```diff
@@ -0,0 +1,42 @@
+        try {
+            ClovaRequest request = ClovaRequest.createClovaRequest(clovaRecommendationRequest);
+
+            return clovaStudioWebClient.post()
+                    .uri(apiUrl)
+                    .bodyValue(request)
+                    .retrieve()
+                    .bodyToMono(ClovaResponse.class)
+                    .block();
+        } catch (WebClientException e) {
```
여기서 block을 걸면 reactive의 이점을 누릴 수 있을까요?

reactive가 항상 빠른것은 아니며, 잘못사용하면 오히려 느려지는 경우가 많은데요. 가장 많이 하는 실수중 하나가, reactive 코드에서 blocking I/O를 유발하는 작업을 할때 인 것 같아요.

reactive가 어떻게 성능적인 이점을 주는지, 알아보셔도 좋을거 같아요.

devxb reviewed (commented):
클린아키텍처, DDD와 같은 개념이 쉽지 않았을거 같은데, 여러가지 시도를 해주신것 같아서 재미있게 봤습니다. 수고하셨습니다 👍 

질문 주신내용중 하나인, `3. 프로젝트의 품질을 향상하고 싶습니다.` 에 대해서는 여기에 답해보도록 할게요.

사실 어떤 방향으로의 품질 향상인지가 약간 추상적이라 생각나는대로 적어보겠습니다 🙇‍♂️ 

- 이미 알고계실수도 있을거 같은데, AggregateRoot, Domain service, Application Service 에 대해서 공부해보셔도 좋을거 같습니다.

- 클린아키텍처가 과연 좋을까요?

- 테스트작성 방법중에서 Describe-Context-It 이라는 패턴이 있습니다. DDD를 하게 되면 도메인이 비대해지면서 테스트코드도 하나의 클래스에 몰려있는 경우가 발생하는데요. 이때, Describe-Context-It을 쓰면 늘어나는 테스트코드를 관리하기 쉬워집니다.
참고: https://johngrib.github.io/wiki/junit5-nested/

- sonar cloud를 쓰면 lint를 어느정도 자동화 해서 관리해줍니다. https://www.sonarsource.com/products/sonarcloud/

- application service를 활용하다보면 모든 로직을 트랜잭션으로 묶지 못하는 경우가 발생하는데요. 이때, 업데이트를 일부만 성공하게 된다면 데이터 정합성은 깨지게 됩니다. 이런 상황을 어떻게 방지할 수 있을까요?

- 각 도메인 끼리는 어떻게 호출을 해야할까요? 

- API가 중복(따닥) 호출되면 어떻게 될까요? 어떻게 멱등한 API를 구현할 수 있을까요?

- 일단 유저를 모으는게 중요한것 같아요. (안된다면 기술을 해야겠지만요) 같은 개선을 하더라도 유저관점에서 유의미한 개선을 했을때 오는 임팩트가 크다고 생각합니다.

더 많을거 같은데 당장 생각나는건 이정도네요 ㅎㅎ

HoyeongJeon committed:
feat: (#65) 빠른시작활동선택의 response dto를 만든다.

HoyeongJeon committed:
feat: (#65) QuickStartActivitySelect는 QuickStartActivityResponse를 반환한다.

HoyeongJeon committed:
fix: (#64) QuickStartActivitySelectSwagger의 중복 PathVariable을 제거한다.

HoyeongJeon committed:
feat: (#64) 추천 재시도 이후 받은 값 중 일부만 올바른 키워드를 가지고 있는 경우, 올바른 키워드를 가지고있는 추천활동만 반환한다.

HoyeongJeon committed:
feat: (#64) userActivityController의 반환값 이름을 수정한다.

HoyeongJeon committed:
refactor: (#64) 불필요한 조건을 제거한다.

HoyeongJeon committed:
feat: (#80) MapService 인터페이스를 생성한다.

HoyeongJeon committed:
feat: (#80) NaverMapErrorCode를 작성한다.

HoyeongJeon committed:
feat: (#80) NaverMapService를 구현한다.

anxi01 committed:
feat: (#81) OpenAI API를 연동한다

anxi01 committed:
feat: (#84) Youtube API를 연동한다

HoyeongJeon committed:
feat: (#80) NaverMapResponse dto를 생성한다.

HoyeongJeon committed:
feat: (#80) NaverMapErrorCode에 파싱 에러를 추가한다.

HoyeongJeon committed:
feat: (#80) NaverMapResponse로 반환값을 변경한다.

HoyeongJeon committed:
refactor: (#80) MapService 를 PlaceInfoProvider로 변경한다.

HoyeongJeon committed:
chore: (#88) email 전송을 위한 의존성을 추가한다.

HoyeongJeon committed:
chore: (#88) MailConfiguration을 작성한다.

HoyeongJeon committed:
feat: (#88) Email 전송을 위한 Util을 구현한다.

HoyeongJeon committed:
feat: (#88) MailErrorCode를 작성한다.

HoyeongJeon committed:
refactor: (#88) EmailUtil의 폴더 구조를 수정하고 예외처리와 로깅을 추가한다.

HoyeongJeon committed:
fix: (#91) AIRecommendationRequest 유효성 검사 도중 발생하는 에러를 ActivityErrorCode로 옮긴다.

HoyeongJeon committed:
fix: (#91) ClovaRecommendationRequest를 AIRecommendationRequest로 수정한다.

HoyeongJeon committed:
fix: (#91) Message의 폴더 위치를 변경한다.

HoyeongJeon committed:
fix: (#91) RecommendationProvider 의 DIP 적용을 위해 Clova와의 결합을 분리한다.

HoyeongJeon committed:
refactor: (#88) Email 입력 값의 유효성 검사를 추가한다.

HoyeongJeon committed:
refactor: (#91) 불필요한 폴더를 삭제한다.

HoyeongJeon committed:
refactor: (#91) 불필요한 개행과 어노테이션을 제거한다.

HoyeongJeon committed:
refactor: (#91) WebClient를 Provider에서 직접 구현한다.

HoyeongJeon committed:
feat: (#88) EmailUtilTest 코드를 작성한다.

HoyeongJeon committed:
refactor: (#91) clova에 종속된 컨트롤러 및 서비스에선 Request dto의 이름을 clovaRecommendationRequest로 사용한다.

anxi01 committed:
refactor: (#78) property 하위 oauth 디렉토리를 생성하여 oauth property를 관리한다

anxi01 committed:
feat: (#78) RabbitMQ 의존성을 추가한다

anxi01 committed:
feat: (#78) RabbitMQ Delayed Message Plugin을 Docker Compose에 추가한다

anxi01 committed:
feat: (#78) ApplicationContext에 등록된 빈을 가져올 수 있다

anxi01 committed:
feat: (#78) RabbitMQ Configuration을 설정한다

anxi01 committed:
feat: (#78) RabbitMQ의 Property, Producer, Consumer 기초 클래스를 생성한다

anxi01 committed:
feat: (#78) 활동의 자투리시간으로 종료 큐 메시지를 발행한다

anxi01 committed:
feat: (#78) 큐에서 활동 종료 메시지를 소비한다

HoyeongJeon committed:
refactor: (#88) SendEmailRequest 인자 이름을 변경한다.

HoyeongJeon committed:
feat: (#87) 월간 활동 요약을 보기위한 request dto를 생성한다.

HoyeongJeon committed:
feat: (#87) 월간 활동 반환을 위한 response dto를 생성한다.

HoyeongJeon committed:
feat: (#87) 월간 활동 반환을 위한 dao를 생성한다.

HoyeongJeon committed:
feat: (#87) 특정 월의 1일부터 마지막날을 반환하는 Util을 생성한다.

HoyeongJeon committed:
feat: (#87) 월간 활동 요약을 제공하는 서비스를 구현한다.

HoyeongJeon committed:
feat: (#87) 월간 활동 요약을 제공하는 컨트롤러를 구현한다.

HoyeongJeon committed:
refactor: (#87) Controller API 이름을 수정한다.

HoyeongJeon committed:
refactor: (#87) Wrapper 타입을 primitive 타입으로 변경한다.

HoyeongJeon committed:
refactor: (#87) api 엔드포인트의 이름을 수정한다.

HoyeongJeon committed:
refactor: (#87) util에 있던 MonthRange를 converter로 분리한다.

HoyeongJeon committed:
refactor: (#87) MonthRange를 삭제하고 변환 메서드를 TimeUtil에 구현한다.

HoyeongJeon committed:
refactor: (#87) ModelAttribute를 삭제한다.

HoyeongJeon committed:
refactor: (#87) Overview 반환 시 유저의 가입월도 보내준다.

HoyeongJeon committed:
feat: (#95) 이번 달 키워드 활동 조회를 위한 request dto와 response dto를 생성한다.

HoyeongJeon committed:
feat: (#95) 이번 달 키워드 활동 조회를 위한 DAO를 구현한다.

HoyeongJeon committed:
feat: (#95) Keyword Category와 이미지를 매핑하는 mapper를 구현한다.

HoyeongJeon committed:
feat: (#95) ReadActivitiesByMemberAndKeywordInMonthService를 구현한다.

HoyeongJeon committed:
feat: (#95) ReadActivitiesByMemberAndKeywordInMonthController를 구현한다.

HoyeongJeon committed:
feat: (#95) 유효성 검사를 Swagger에 위임하고, PathVariable을 사용한다.

HoyeongJeon committed:
chore: (#98) build.gradle에 Sentry 설정을 추가한다.

HoyeongJeon committed:
chore: (#98) 예외 발생 시 Sentry에 기록한다.

anxi01 committed:
feat: (#90) 활동 캘린더 조회 Query를 DAO에 작성한다

anxi01 committed:
feat: (#90) 활동 캘린더 조회 API를 추가한다

anxi01 committed:
feat: (#90) 활동 캘린더 조회 API 스웨거를 작성한다

anxi01 committed:
fix: (#103) 활동 자동 종료 서비스를 테스트코드에서 Mock 처리한다

HoyeongJeon committed:
fix: (#105) build.gradle 의 SENTRY AUTH TOKEN을 제거한다.

anxi01 committed:
feat: (#102) 키워드 카테고리의 description으로 카테고리 Enum 값을 가져올 수 있다

anxi01 committed:
fix: (#102) 활동 타입의 description이 아닌 Enum 값으로 UserMessage를 생성하도록 변경한다

anxi01 committed:
feat: (#102) OpenAI 추천 프롬프트를 작성한다

anxi01 committed:
feat: (#102) OpenAI ErrorCode를 생성한다

anxi01 committed:
feat: (#102) OpenAI를 이용한 온라인, 온오프라인 추천을 할 수 있다

anxi01 committed:
feat: (#102) OpenAI 추천 API 스웨거를 작성한다

HoyeongJeon committed:
refactor: (#87) startDayOfMonth, endDayOfMonth를 변수로 뺀다.

HoyeongJeon committed:
refactor: (#87) swagger 엔드포인트를 변경한다.

HoyeongJeon committed:
refactor: (#87) 월간활동 오버뷰 조회 시 가입년도도 함께 반환한다.

HoyeongJeon committed:
refactor: (#87) 원시형의 Null 검사를 제거한다.

HoyeongJeon committed:
refactor: (#87) MonthlySavedTimeAndActivityCountResponse의 반환값을 원시형으로 변경한다.

HoyeongJeon committed:
refactor: (#87) 코드 내 불필요한 공백을 제거한다.

HoyeongJeon committed:
refactor: (#95) 래퍼 클래스를 primitive 로 변환한다.

HoyeongJeon committed:
refactor: (#95) Swagger 설명을 변경한다.

anxi01 committed:
feat: (#102) 키워드를 통해 유튜브 비디오 URL을 검색할 수 있다

anxi01 committed:
fix: (#102) OpenAI 추천을 얻는 방식을 동기식 방식으로 변경한다

HoyeongJeon committed:
refactor: (#99) 클로바 스튜디오 프롬프트를 수정한다.

HoyeongJeon committed:
refactor: (#99) PlaceInfoProvider의 반환형을 제네릭으로 수정한다.

HoyeongJeon committed:
feat: (#99) KakaoPlaceInfoProvider를 구현한다.

HoyeongJeon committed:
feat: (#99) KakaoMapResponse dto를 생성한다.

HoyeongJeon committed:
feat: (#99) KakaoMapErrorCode를 생성한다.

HoyeongJeon committed:
refactor: (#99) ClovaRecommendationResponse dto에 좌표와 위치url을 추가한다.

HoyeongJeon committed:
refactor: (#99) GetRecommendationsFromClovaService에서 Kakao 위치 검색을 사용하도록 변경한다.

HoyeongJeon committed:
refactor: (#95) 변수명을 수정한다.

anxi01 committed:
fix: (#102) OpenAI 추천 API에 검증 어노테이션을 추가한다

anxi01 committed:
feat: (#112) 백엔드 배포 도메인을 CORS Origin에 추가한다

HoyeongJeon committed:
refactor: (#95) ReadActivitiesByMemberAndKeywordInMonthController 엔드포인트를 PathVariable에서 쿼리스트링을 사용하도록 변경한다.

HoyeongJeon committed:
refactor: (#95) ActivitiesByMemberAndKeywordInMonthResponse의 반환 순서를 변경한다.

HoyeongJeon committed:
refactor: (#95) ActivityWithTitleAndSavedTimeResponse 반환 변수 이름을 변경한다.

HoyeongJeon committed:
refactor: (#95) 분리되어 있는 조회쿼리를 하나로 합친다.

HoyeongJeon committed:
refactor: (#95) keywordImageMapper를 Keyword 안에 스태틱 메서드로 정의한다.

HoyeongJeon committed:
refactor: (#95) KakaoPlaceInfoProvider의 url을 프로퍼티로 관리한다.

HoyeongJeon committed:
refactor: (#99) KakaoMapResponse 변수를 카멜케이스로 변환한다.

anxi01 committed:
feat: (#109) Spring Batch 의존성을 추가한다

anxi01 committed:
fix: (#109) 이메일 수신자를 여러명으로 변경한다

anxi01 committed:
feat: (#109) 시간 범위 내 빠른 시작을 사용자별 1개씩 가져오는 JPQL을 추가한다

anxi01 committed:
feat: (#109) Batch를 이용해서 15분 주기로 빠른시작 알림 메일을 전송한다

anxi01 committed:
feat: (#109) Scheduler를 이용해서 15분 주기로 빠른시작 알림 메일을 전송한다

HoyeongJeon committed:
refactor: (#110) AuthorizationInterceptor에서 토큰 추출을 쿠키에서 하도록 변경한다.

HoyeongJeon committed:
refactor: (#110) LoginMemberArgumentResolver에서 토큰 검증 시 쿠키에서 토큰을 뽑아내도록 변경한다.

HoyeongJeon committed:
refactor: (#110) 로그인 성공 시 쿠키에 토큰과 유저정보를 넣어준다.

HoyeongJeon committed:
refactor: (#110) AuthenticationErrorCode를 수정한다.

HoyeongJeon committed:
refactor: (#110) 토큰 재발급시 새로운 accessToken을 쿠키에 넣어준다.

HoyeongJeon committed:
refactor: (#110) LoginResponse에서 refreshToken을 제거한다.

HoyeongJeon committed:
refactor: (#110) 로그인 시 refreshToken을 redis에 저장한다.

HoyeongJeon committed:
refactor: (#110) 만료된 accessToken에서 memberId를 추출하는 메서드를 추가한다.

HoyeongJeon committed:
refactor: (#110) refreshToken을 가져올 때 검증 로직을 추가한다.

HoyeongJeon committed:
feat: (#110) logout 컨트롤러를 추가한다.

HoyeongJeon committed:
feat: (#110) 인증 우회 uri를 설정한다.

HoyeongJeon committed:
feat: (#110) Swagger 설정을 cookie를 사용하도록 변경한다.

HoyeongJeon committed:
feat: (#110) 토큰 재발급, 로그아웃 컨트롤러에 Swagger를 설정한다.

HoyeongJeon committed:
feat: (#114) activityType이 ONLINE_AND_OFFLINE인 경우 응답을 3개 반환한다.

HoyeongJeon committed:
refactor: (#114) String 값과 Keyword.Category를 비교하는 스위치문을 Enum 사용으로 바꾼다.

HoyeongJeon committed:
refactor: (#110) Swagger-ui로 요청을 보낼 경우엔 Authorization Header를 통해 인증/인가를 한다.

anxi01 committed:
feat: (#118) 온라인, 오프라인 추천 API를 하나로 통합한다

anxi01 committed:
fix: (#120) 수신자를 여러 명으로 처리하도록 SendEmailRequest를 변경하면서 테스트 코드도 수정한다

HoyeongJeon committed:
refactor: (#123) ClovaStudioPrompt를 수정한다.

HoyeongJeon committed:
refactor: (#123) Keyword Category와 ImageUrl 매핑을 상수로 만든다.

HoyeongJeon committed:
refactor: (#123) 키워드 카테고리로 반환하던 응답을 키워드로 반환한다.

anxi01 committed:
fix: (#127) 중복된 닉네임일 때 에러 처리가 아닌 DTO에 중복 결과를 반환한다

anxi01 committed:
fix: (#129) 중복 닉네임 검증에서 반환 값이 변경되면서 테스트 코드를 수정한다

HoyeongJeon committed:
feat: (#126) Member 엔티티에 프로필 수정 메서드를 구현한다.

HoyeongJeon committed:
feat: (#126) 프로필 수정 스웨거와 컨트롤러를 생성한다.

HoyeongJeon committed:
feat: (#126) 프로필 수정 리퀘스트 DTO를 생성한다.

HoyeongJeon committed:
feat: (#126) 프로필 수정 서비스를 구현한다.

HoyeongJeon committed:
feat: (#126) 프로필 수정 서비스 테스트를 생성한다.

anxi01 committed:
feat: (#124) Member 도메인 이메일 알림 변경 도메인 메소드를 추가한다

anxi01 committed:
feat: (#124) 이메일 알림 설정 API를 추가한다

anxi01 committed:
feat: (#124) 이메일 알림 설정 API 스웨거를 작성한다

anxi01 committed:
feat: (#124) 이메일 알림 설정 테스트코드를 작성한다

HoyeongJeon committed:
refactor: (#126) EditMemberProfileSwagger에 MemberErrorCode를 추가한다.

HoyeongJeon committed:
feat: (#126) 닉네임 유효성 검사를 ValidateNicknameService에 위임한다.

HoyeongJeon committed:
feat: (#126) EditMemberProfileService edit 메서드에 @Transactional을 추가한다.

HoyeongJeon committed:
feat: (#126) EditMemberProfile의 반환값을 void로 변경한다.

HoyeongJeon committed:
feat: (#126) EditMemberProfile의 반환값을 void로 변경한다.

anxi01 committed:
feat: (#133) 이미지 url을 관리하는 Property를 추가한다

anxi01 committed:
feat: (#133) Category로 Image를 변환하는 Converter를 추가한다

HoyeongJeon committed:
refactor: (#126) EditMemberProfileService의 불필요한 검증 로직을 제거한다.

anxi01 committed:
fix: (#133) ImageConverter를 사용해서 이전 코드를 수정한다

anxi01 committed:
feat: (#133) ImageConverter 테스트 코드를 작성한다

anxi01 committed:
fix: (#135) 로그인 응답에 사용자 가입 날짜를 추가한다

anxi01 committed:
feat: (#135) 온보딩 응답에 사용자 정보를 추가한다

anxi01 committed:
fix: (#135) 온보딩 가입 테스트 코드를 수정한다

HoyeongJeon committed:
chore: (#86) HealthController를 수정한다.

HoyeongJeon committed:
chore: (#86) PR 이벤트가 발생했을 경우, github.sha가 pr 이벤트의 해시값을 사용하도록 변경한다.

anxi01 committed:
fix: (#138) Keyword 밸류 타입 POJO화 및 @Embeddable로 인프라 레이어에 추가한다

HoyeongJeon committed:
chore: (#86) 배포 서버의 deploy.sha 에 환경변수를 설정한다.

anxi01 committed:
fix: (#139) 배포 서버 JVM의 타임존을 Asia/Seoul으로 변경한다

anxi01 committed:
fix: (#142) 로그인 시 사용자 정보를 넘겨준다

HoyeongJeon committed:
feat: (#144) 로그아웃을 구현한다.

HoyeongJeon committed:
feat: (#144) 멤버 프로필 조회 Response dto를 생성한다.

HoyeongJeon committed:
feat: (#144) 멤버 프로필 조회 컨트롤러와 스웨거를 작성한다.

HoyeongJeon committed:
feat: (#144) 멤버 프로필 조회 테스트를 작성한다.

anxi01 committed:
fix: (#146) 로그인 시 온보딩을 진행하지 않은 사용자의 닉네임은 null로 반환한다

anxi01 committed:
feat: (#146) Role이 GUEST인 경우 닉네임은 null로 반환하는 테스트 코드를 작성한다

anxi01 committed:
refactor: (#148) 멤버 생성 서비스에서 함수를 분리하고 @Transactional를 붙여 가독성과 안정성을 높일 수 있다

HoyeongJeon committed:
refactor: (#150) MemberID 생성시 Sequential UUID를 사용하도록 변경한다.

anxi01 committed:
feat: (#151) Thymeleaf 의존성을 추가한다

anxi01 committed:
fix: (#151) 메일 전송시 HTML을 본문에 넣기위해 MimeMessage로 변경한다

anxi01 committed:
feat: (#151) 빠른 시작 메일 전송 템플릿을 적용한다

anxi01 committed:
fix: (#153) 메인 화면에서 현재 시간에 가장 가까운 빠른 시작을 가져오지 못하는 이슈 대응

HoyeongJeon committed:
refactor: (#155) Clova Prompt를 수정한다.

HoyeongJeon committed:
refactor: (#155) Keyword 에서 NATURE를 삭제한다.

HoyeongJeon committed:
refactor: (#155) Keyword 관련 로직을 수정한다.
- ONLINE_AND_OFFLINE 시 OFFLINE 요청에 SOCIAL이 들어올 경우 걸러낸다.
- ClovaStudio 응답에서 파싱용 PATTERN을 수정한다.
- 응답의 키워드가 여러개일 경우, 키워드를 분리하고 랜덤한 하나의 키워드만 전달한다.

HoyeongJeon committed:
refactor: (#150) SequentialUUID의 위치를 변경한다.

anxi01 committed:
refactor: (#151) 이메일 전송 스케쥴러 및 배치 메소드를 분리한다

anxi01 committed:
fix: (#151) 이메일 수신자를 벌크 단위에서 단건으로 변경한다

anxi01 committed:
fix: (#151) 빠른 시작의 시간대를 구하는 쿼리에서 중복 데이터를 제거한다

anxi01 committed:
feat: (#151) 이메일을 비동기적으로 전송한다

anxi01 committed:
fix: (#151) 이메일 수신자 쿼리를 멤버 도메인에서 조회하도록 변경한다

anxi01 committed:
feat: (#151) 사용자와 빠른 시작 정보를 가지고 메일 제목을 커스텀할 수 있다

anxi01 committed:
fix: (#157) 활동 키워드에서 키워드 이미지는 투명도 30으로 변경한다

HoyeongJeon committed:
fix: (#155) 사용하지 않는 에러코드를 삭제한다.

HoyeongJeon committed:
fix: (#155) NATURE 키워드를 삭제한다.

HoyeongJeon committed:
fix: (#155) new Random을 final로 선언하고, parsedKeywordText 내 추가적인 trim()으로 의도치않은 공백을 제거한다.

HoyeongJeon committed:
fix: (#155) static 키워드를 추가한다.

anxi01 committed:
fix: (#159) 조각 별 상세 정보 페이지에서는 원본 키워드 이미지를 반환한다

anxi01 committed:
fix: (#160) 조각 별 상세 정보가 아닌 활동 키워드에서 키워드 이미지는 투명도 30으로 변경한다

anxi01 committed:
fix: (#162) 수신자를 단 건으로 처리하도록 SendEmailRequest를 변경하면서 테스트 코드도 수정한다

anxi01 committed:
refactor: (#164) 빠른시작 도메인 분리 및 응용서비스를 도메인 서비스로 이관한다

anxi01 committed:
refactor: (#164) 활동 도메인 응용서비스를 도메인 서비스로 이관한다

anxi01 committed:
refactor: (#164) 멤버 도메인 응용서비스를 도메인 서비스로 이관한다

anxi01 committed:
refactor: (#164) 도메인 분리 및 서비스 이관으로 인한 테스트코드 패키지를 변경한다

HoyeongJeon committed:
fix: (#166) 사용하지 않는 테스트를 삭제한다.

HoyeongJeon committed:
refactor: (#168) gradle 파일을 역할별로 분리한다.

HoyeongJeon committed:
refactor: (#170) Recommendation 도메인의 dto 구조를 controller 사용 여부에 따라 재구성한다.

HoyeongJeon committed:
refactor: (#170) QuickStart 도메인의 dto 구조를 controller 사용 여부에 따라 재구성한다.

HoyeongJeon committed:
refactor: (#170) Member 도메인의 dto 구조를 controller 사용 여부에 따라 재구성한다.

HoyeongJeon committed:
refactor: (#170) Auth 도메인의 dto 구조를 controller 사용 여부에 따라 재구성한다.

HoyeongJeon committed:
refactor: (#170) Activity 도메인의 dto 구조를 controller 사용 여부에 따라 재구성한다.

anxi01 committed:
docs: README를 작성한다

anxi01 committed:
fix: (#174) 온보딩에서 닉네임에 초성 입력이 가능하다

anxi01 committed:
fix: (#174) 빠른시작 생성시 초성 입력이 가능하다

HoyeongJeon committed:
fix: (#176) 추천 활동 장소의 url이 없는 경우, 카카오맵 검색으로 넘긴다.

HoyeongJeon committed:
fix: (#178) Prompt를 수정한다.

HoyeongJeon committed:
refactor: (#180) Prompt를 수정한다.

HoyeongJeon committed:
refactor: (#180) 불필요한 코드를 메서드 삭제 및 텍스트 값에 일치하는 카테고리 변환 로직을 리팩토링한다.

HoyeongJeon committed:
refactor: (#110) 사용하지 않는 LoginResponse를 삭제한다.

HoyeongJeon committed:
refactor: (#110) http://localhost:5173를 cors 허용한다.

HoyeongJeon committed:
refactor: (#110) Dao 반환 값을 래퍼 클래스가 아닌 원시형으로 변경한다.

HoyeongJeon committed:
refactor: (#110) refreshToken 발급 시 claim 속 유저 정보를 제거한다.

HoyeongJeon committed:
refactor: (#110) refreshToken 발급 시 토큰 저장소에 refreshToken:memberId의 형태로 저장한다.

HoyeongJeon committed:
refactor: (#110) 토큰 재발급 로직을 변경한다.

HoyeongJeon committed:
refactor: (#110) 토큰 재발급 시 쿠키에 있는 refreshToken을 사용하고, Swagger를 변경한다.

HoyeongJeon committed:
refactor: (#110) 로그인 시 쿠키에 AccessToken과 RefreshToken을 넣는다.

HoyeongJeon committed:
refactor: (#110) 로그아웃 시 쿠키를 초기화하고, 토큰 저장소에 있는 RefreshToken을 삭제한다.

HoyeongJeon committed:
refactor: (#110) RefreshTokenServiceTest를 수정한다.

HoyeongJeon committed:
refactor: (#110) LoginMemberArgumentResolverTest를 수정한다.

HoyeongJeon committed:
chore: (#110) 테스트를 위해 pr 시 ci/cd가 돌아가도록 변경한다.

HoyeongJeon committed:
fix: 쿠키에 httpOnly 설정을 true로 변경한다.

HoyeongJeon committed:
fix: 배포 스크립트의 pull_requests 시 배포되는 부분을 제거한다.

HoyeongJeon committed:
refactor: AuthenticationErrorCode의 NOT_EXIST_TOKEN의 이름을 NOT_EXIST_TOKEN_COOKIE로 변경한다.

HoyeongJeon committed:
refactor: 불필요한 주석을 제거한다.

HoyeongJeon committed:
refactor: 불필요한 반환 값을 제거한다.

HoyeongJeon committed:
refactor: enum 상수 비교를 equals()에서 == 수정한다.(NPE 발생 이슈)

HoyeongJeon committed:
refactor: 코드 컨벤션을 지킨다.

HoyeongJeon committed:
refactor: 로그인 리턴값에 대한 새로운 Response Record를 생성한다.

HoyeongJeon committed:
chore: geoip2 의존성을 생성한다.

HoyeongJeon committed:
chore: GeoIpConfiguration을 설정한다.

HoyeongJeon committed:
feat: 두 좌표 사이 거리 계산을 위한 유틸을 생성한다.

HoyeongJeon committed:
feat: 로그인 한 유저의 IP 획득용 ArgumentResolver를 생성한다.

HoyeongJeon committed:
feat: argumentResolver를 추가한다.

HoyeongJeon committed:
feat: GeoLocationService를 구현한다.

HoyeongJeon committed:
chore: gitignore를 수정한다.

HoyeongJeon committed:
feat: refreshToken 저장 시 ip를 함께 저장한다.

HoyeongJeon committed:
feat: 로그인 시 ip를 함께 받아온다.

HoyeongJeon committed:
feat: 토큰 재발급 시 RTR을 적용한다.

HoyeongJeon committed:
feat: 토큰 재발급 시 100 km 이상에서 들어온 요청에 대해 재발급을 반려한다.

HoyeongJeon committed:
feat: Redis 토큰 저장 메서드를 변경한다.

HoyeongJeon committed:
feat: 100km 이상 떨어진 곳에서의 요청 거절 에러를 만든다.

HoyeongJeon committed:
feat: 토큰 반환 dto 내용을 변경한다.

HoyeongJeon committed:
feat: RotateTokenSwagger 및 Redis 키-밸류 전체 삭제 메서드를 구현한다.

HoyeongJeon committed:
feat: 테스트를 작성한다.

HoyeongJeon committed:
chore: maxmind db를 .gitignore에 추가한다.

HoyeongJeon committed:
chore: ci/cd 스크립트를 수정한다.

HoyeongJeon committed:
refactor: 에러 메시지를 수정한다.

HoyeongJeon committed:
feat: 토큰 재발급 유효 거리를 환경변수로 관리한다.

HoyeongJeon committed:
feat: 쿠키 발급 시 secure, sameSite를 추가한다.
